File: 2019/haskell/Day5.hs
module Main where

import Control.Monad.State

import Intcode

main :: IO ()
main = do mach <- readIntcode "../data/day5"
          let resulta = evalState (run [1]) mach
          let (_,rest) = span (==0) resulta
          case rest of
            [a] -> putStrLn $ "part a: " ++ show a
            nzs -> putStrLn $ "part a: failure: non-zero output was " ++ show nzs

          let resultb = evalState (run [5]) mach
          case resultb of
            [b] -> putStrLn $ "part b: " ++ show b
            _ -> putStrLn $ "part b: failure: non-singleton output was " ++ show resultb

File: 2019/rust/src/bin/day5.rs
use aoc2019::intcode::*;

fn main() {
    let mem = read_intcode("../data/day5");
    let mut macha = IC { ip: 0, mem };
    let mut machb = macha.clone();

    let input = vec![1_i32];
    let mut input = input.iter().map(|x| *x);
    let mut out = macha.run(&mut input).skip_while(|x| *x == 0_i32);
    match out.next() {
        Some(out1) => match out.next() {
            None => println!("part a: {}", out1),
            Some(b) => println!(
                "part a, tests FAILED, had some zeros and then {},{},{:?}",
                out1,
                b,
                out.collect::<Vec<_>>()
            ),
        },
        None => println!("part a, unknown failure, we only output zeros"),
    }

    let input = vec![5_i32];
    let mut input = input.iter().map(|x| *x);
    let mut out = machb.run(&mut input);
    match out.next() {
        Some(res) => match out.next() {
            None => println!("part b: {}", res),
            Some(b) => println!(
                "part b, failure, multiple outputs: {},{},{:?}",
                res,
                b,
                out.collect::<Vec<_>>()
            ),
        },
        None => println!("part b, unknown failure, no output"),
    }
}
